### DISCUSSION TITLE
Screen Mirroring Resize

### DISCUSSION THREAD

**Discussion by marshyrob** [BimmerGestalt/AAIdrive#517] [Q&A]
Great work on this app, it really is developing and taking shape.  

I recently tried the screen mirror addon and it, like the the standard BMW screen mirror functionality, only uses a small area of the screen to mirror the phone.  I know this is to do with the way miracast works, ie the aspect ratio's are different so its wont use the whole screen.  But i was wondering if there is anyway to implement some sort of logic that can make the mirroring use the correct aspect ratio of the screen its mirroring to, thus using more of the real estate avaialble to it.  I currently find BMW mirroring pointless as it is actually dispalying on the BMW screen smaller than the screen of the phone.  I may aswell just balance the phone over the car's screen and have a better experience. 

Not sure its possible but thought id ask.

**Comment by hufman** [comment]
Thank you for the kind words!
Internally, the screen mirroring addon creates a virtual display of the same resolution as the car display, and Android fits the mirrored content into this virtual display including adding black bars as necessary. I don't know of any way to change that default behavior, either forcing it to crop off edges or stretching without preserving the ratio instead of shrinking to fit. However, this part is regular Android programming, and so I would welcome contributions from any Android dev experts!
What exact behavior would you like to see?

**Reply by marshyrob** [reply]
Yes i assumed that would be the case, its a shame as its not really that useful due to the size it renders on the car display, its easier to look at my phone!  if it could be full screen keeping the aspect ratio it would be more useful (Google Maps, Movies etc) but as it is i cant really find a use for it.  Its the same with the BMW official implementation, so its no worse, just the limitation of mirroring i suppose!  Thanks for responding, keep up the good work.

**Reply by Jimbo757767** [reply]
I've got the wider, split screen on ID6 and like to run google maps on the main part and the car's nav on the RH split. I find with this that the split cuts off about a third of the RHS of the screen mirroring of google maps. Is it possible to do anything about that?

**Reply by hufman** [reply]
I haven't found a way to detect whether the car is running in split screen mode or not. For the maps functionality, I have a checkbox for the user to configure splitscreen mode or not. Since the screen mirroring addon is a proof-of-concept, I haven't found time to implement a more advanced settings UI or develop partial screen mirroring.

**Reply by Jimbo757767** [reply]
Thanks. You mention a user checkbox for split screen. I can't Find it. I'm using version 1.3.2.

**Reply by hufman** [reply]
It's in the Map tab, which is currently an optional compile-time feature.

**Reply by Jimbo757767** [reply]
Ok, if I compile that will it work only for the map tab, or for screen mirroring as well?

**Reply by hufman** [reply]
It will only apply to the map. The addons are separate apps and don't share any settings with AAIdrive.

**Reply by patalmqx** [reply]
For curiosity: do you have some photos of the function or better short video demonstrating the function? Resolution, update rate, screen  usage..

**Reply by hufman** [reply]
A video of my phone (I forget if it was the Galaxy S9 or the Pixel 2) streaming to my Mini over USB can be found [here](https://twitter.com/snes_rocks/status/1441946229718487050). It records at the [resolution of the car minus padding](https://github.com/BimmerGestalt/IDriveConnectKit/blob/main/src/main/java/io/bimmergestalt/idriveconnectkit/RHMIDimensions.kt#L56), the bitrate depends on how busy the screen is for the JPG encoder ([quality 60](https://github.com/BimmerGestalt/IDriveConnectAddons/blob/1fd3316b41053b9cb237797f4aabd94f149e20b7/screen_mirror/src/main/java/io/bimmergestalt/idriveconnectaddons/screenmirror/ScreenMirrorProvider.kt#L39) over USB, [quality 30](https://github.com/BimmerGestalt/IDriveConnectAddons/blob/1fd3316b41053b9cb237797f4aabd94f149e20b7/screen_mirror/src/main/java/io/bimmergestalt/idriveconnectaddons/screenmirror/CarAppService.kt#L72) over BT), and the framerate goes as fast as the transport can handle (up to the phone screen's actual update screen). The update rate is capped at 0.5hz while in motion, unless Android Auto is running. The screen usage is based on the aspect ratios of fitting the phone's display into the virtual display, so some very tall phones will have bars along the top and bottom.

**Comment by patalmqx** [comment]
Perfect as always. Maybe link the vid on a more prominent page at the
github Portal?

How to understand  Android Auto need for mot having max. 0 5Hz framerate?
Nbt e.g. doesnt support AA. Or you mean phone side? I also researched max.
0,5 Hz in your application due to technical limitation. Your vid is much
more fluent..

My usecase: gmaps!

Walter Huf ***@***.***> schrieb am So., 20. Feb. 2022, 22:47:

**Reply by hufman** [reply]
The 0.5hz update rate is to reduce distracted driving, by not making it easy to (for example) watch Youtube while driving. I picked 0.5hz as a compromise to allow specifically Google Maps to work well. When AAIdrive detects that the phone is displaying Android Auto, it trusts Android Auto's distracted driving mitigations and so AAIdrive won't artificially reduce the framerate. The linked video shows a faster update because the car is parked. Bluetooth will also go a lot slower than the USB used in that video.

**Reply by patalmqx** [reply]
But i could use my own parameter to compile the addon to have full update rate right?

0,5Hz is quite slow for following complicated turn sequences.

What do you mean by android auto? I guess my id4 nbt doesnt support it although my phone would.. you mean some other kind of android car mode?

**Reply by hufman** [reply]
Indeed, it's open source under the MIT license and so you can copy the code and do whatever you want to it :)
Yes exactly, if you open Android Auto on your phone, it will take over the screen and show a car-optimized version of Google Maps on your phone screen, and this mode will be allowed to update through Screen Mirroring as fast as possible. However, that will still be only about 1-4fps over Bluetooth bandwidth.

**Reply by patalmqx** [reply]
I tried it now an it works! Just the size is far away from fullscreen. Is it due to a technical limit in id4 6,5" cars? In your id5/6 video the additional 'dead space' appears to be way smaller. Or is it due to the large vs  small screen difference? @hufman 
![20220306_123051](https://user-images.githubusercontent.com/86716143/156924825-08df8206-9b6c-4bb8-83ff-5515333713ab.jpg)

**Reply by hufman** [reply]
Because the screen mirroring is just another app in the car, there are several constraints on the space available to show the image. A car app can apply a negative padding to move the image over to the left, but then other parts of the car UI will draw over the top of the image. The ID4 version seems to have a bigger control area on the left compared to the ID5+ version. The map module can do this adjustment, because it is more confident that there won't be any important information at that part of the image, but the generic screen mirroring can't.
The blank space above the top and bottom of the image is due to the screen ratio mismatch between the available space in the car and your phone's screen, a mismatch which is greater on the 6.5" screen compared to 8.8" screen.

**Reply by patalmqx** [reply]
So there is no way to escape this apps frame for id4 or stretch the oicture on top of the round inicator ui at the left side of the screen? Mapbox uses the same frame i tested yesterday.

Edit found  my answer. https://github.com/BimmerGestalt/AAIdrive/issues/538#issuecomment-1050026485

**Reply by dimakaob** [reply]
> I tried it now an it works! Just the size is far away from fullscreen. Is it due to a technical limit in id4 6,5" cars? In your id5/6 video the additional 'dead space' appears to be way smaller. Or is it due to the large vs small screen difference? @hufman ![20220306_123051](https://user-images.githubusercontent.com/86716143/156924825-08df8206-9b6c-4bb8-83ff-5515333713ab.jpg)

Can you tell me please, how you managed to open Android Auto on your phone? As i know it's not possible to do it like any other app. Only if the car supports it..

**Reply by hufman** [reply]
My phone has an Android Auto app (version 7.3.620524-release installed from the [Play Store](https://play.google.com/store/apps/details?id=com.google.android.projection.gearhead)), and when I open it I get a little version of the Android Auto interface showing on my phone. What happens when you open yours?

**Reply by bogdan-calapod** [reply]
@hufman does your phone show `Android Auto` in the app list ? 

When I tried to install the app from the link you sent, it says that `This item isn't available in your country`.

![image](https://user-images.githubusercontent.com/5462727/157475015-7d214928-3961-4706-af49-08588141c96c.png)

I'll try playing around with the APK and see how that goes. :)

**Reply by hufman** [reply]
Yes, both my Pixel2 and Galaxy S9 have it in the apps list:
![photo5568821808669501358](https://user-images.githubusercontent.com/1592375/157491410-1f2fe977-e608-476a-9588-5631398ceed7.jpg)

**Comment by patalmqx** [comment]
Thank you for the response!

Even my not car interested wife found it cool to be possible on this car :D

-So the image would never be on top of the UI in ID4? In case of the
mirroring I do not see any significant control ui buttons except the large
'selection' button on the left. Can i upscale and pan the image via the
parameters beyond the app frame?
-Could the maps environment be used to display the 'mirroring pictures'
technically? From what I saw the screen is optimally used.
-The black bars on top/bottom are just inevitable but that is expected!

Maybe i have to learn app development but my skills end at Matlab. Got to
find some time :○


Walter Huf ***@***.***> schrieb am So., 6. März 2022, 17:27:

**Comment by dimakaob** [comment]
I have the same version.
But I can't even open it, because I don't have it in my app lists. As I
know, it opens automatically when you connect it to the car. But we dont
have AA support in our cars..

ср, 9 мар. 2022 г., 18:09 Walter Huf ***@***.***>:

**Comment by dimakaob** [comment]
I guess you have Android 9 or 10 on your phones..

ср, 9 мар. 2022 г. в 19:58, Walter Huf ***@***.***>:

**Comment by magamess** [comment]
Hi @hufman , following the discussion I raised on #538, would it be possible to get a modified apk version for resizing the screen mirroring app to minimize the black margins in a 1280x480 screen?

Happy to test it on my car and provide some feedback!

Thanks!

**Reply by hufman** [reply]
The latest version now tries to fill the screen as much as possible, even if it goes underneath the BMW components.

**Reply by marshyrob** [reply]
how do i get the latest version still says Oct 21 is the latest in the applications section for screen mirroring, can only see the code on the version from 18 hours ago but no .APK

**Reply by bogdan-calapod** [reply]
@marshyrob you'll need to download the latest screen mirror apk [from here](https://github.com/BimmerGestalt/IDriveConnectAddons/suites/5692372688/artifacts/187288683)

**Reply by hufman** [reply]
For some reason it didn't change the release date, but I decompiled the APK in that page and it does include the changes!

**Reply by Krakuanen** [reply]
Full screen mirroring works perfectly on my 2018 Mini Cooper. Amazing!

**Reply by spykee2005** [reply]
Working great... thx!!!

**Reply by patalmqx** [reply]
Is mirroring when using google maps giving reasonable visibility of fetails? Do you habe some pics? 

On my 6 5" screenthe details are rather small, an upgrade to 8 8" screen might be interesting..

**Comment by bpowder1966** [comment]
I keep getting this error message when trying to install the updated screen
mirroring

**Reply by hufman** [reply]
Images don't make it through the email reply and so the error message isn't visible, please attach it directly on GitHub.

**Reply by bpowder1966** [reply]
![SmartSelect_20220320-104234_Package installer](https://user-images.githubusercontent.com/83174195/159177908-b2d22a50-8563-4c4c-a29c-e388a80b8803.jpg)

How do I uninstall the previous version of that is what is required?  I don't see a way to do that

**Reply by bogdan-calapod** [reply]
Go to the Settings -> Apps and look for the `Screen Mirroring` app and you can uninstall it from there

**Reply by hufman** [reply]
You can also tap the checkboxes next to the app in the Addons window to go right to that App Info screen, which should provide the usual Uninstall/ForceStop options

**Comment by marshyrob** [comment]
I'm guessing you get not installed?. If so you need to uninstall the old 
version of screen mirroring before you install the new one.

**Comment by bpowder1966** [comment]
I'm sure this has been asked but there is no way to turn off the screen
during mirroring?

I'd like to get Android auto running to show Google maps and turn off my
phone screen..

**Reply by hufman** [reply]
No way as far as I know!

**Reply by MLiv79** [reply]
> No way as far as I know!

Hi hufman and thanks again for the great app. I'm on nbt evo 6 and I've 2 main issues with screen mirroring. First one as many here the screen don't fit vertically my car screen, second is very laggy, I don't know if it's the best possible or something wrong in my configuration. My car have also wifi connection but i think the app is working only over Bluetooth right?

**Reply by hufman** [reply]
Yes that's correct, the BMW Apps protocol only functions over Bluetooth, and can't use the wifi connection, so there is very little to be done to improve performance. USB works for some people, it's buggy with MyBMW, seems to not work at all in Android 15, and impacts the music app integration.
The screen size issue is related to trying to fit your entire phone screen into the drawable surface of the car app, your phone is probably just too big unfortunately.
